debugger.md

```diff
@@ -0,0 +1,160 @@
+# Debugger proposal
+
+## Background
+
+One of the greatest advantages of being a log based workflow engine is that Temporal workflows are replayable.
+
+A developer can download workflow history and run it through the SDK replayer to materialize the workflow program's
+state.
+
+Workflows, being "just code", can be debugged with tools the developer is familiar with.
+
+## Features
+
+### Overview
+
+Typically, when looking at workflow history, it's clear which workflow task a failure occurs in but it's not immediately
+obvious which line of code that correlates to.
+
+This proposal discusses how we can make the debugging experience better for developers.
+
+### Supported languages
+
+Initially we will support Core based languages (TypeScript and Python) but Go and Java could also be supported with
+additional effort.
+
+The main advantage of going for Core based languages is that those language SDKs run Rust Core in a separate thread
+pool, allowing interaction with the runtime while a workflow is paused by a debugger.
+
+Targeting Core based SDKs will allow us to reuse a lot of the code for Ruby, .NET and possibly other SDKs if we decide
+to rewrite them to be backed by Core.
+
+### Supported IDEs
+
+Initially we will support Microsoft's Visual Studio Code only while keeping in mind other IDEs.
+
+The rationale being that at least for JavaScript VSCode is by far the most popular IDE and we'd get the most value
+investing in providing good DX for that environment.
+
+### Components
+
+- VSCode plugin
+  - Main panel / Workflow history panel
+  - Debug adapter
+- Core replay debug server
+- Lang middleware
+
+#### Main panel
+
+Has input to enter workflow ID and optional run ID with submit button to start a debug session.
```

Comment:
> submit button to start a debug session.

Can you clarify whether this starts a vscode debug session just like it would without the Temporal extension? I just want to confirm that this:

* Starts a replayer w/ deadlock detection off to do normal code stepping
* This main panel is just a easy interface to a launch.json they could otherwise setup
* The history panel is just a panel like call stack and locals that is just giving more info (and allows some stepping at an event/task level)

Reply:
Confirmed.
I'll add clarifications to the proposal.

Reply:
It's currently not possible for Core based SDKs to correlate history events with activation jobs.
@Sushisource we should probably add that ability.